teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+
+- Achieve SLSA L3 supply chain security through Tekton `PipelineRuns`
+- End-to-end trust of the provenance produced by Tekton Chains for Tekton `PipelineRuns`
+- `Tasks` to emit a list of artifacts of unknown length.
+
+### Requirements
+
+- Enable `Tasks` to be able to surface Artifact metadata as an array.
+
+## Proposal
+
+Define the structure of the Artifact metadata that the `Step` needs to write out. Define the file path and variable replacement syntax for the location of the metadata file.
+
+The following chapters in the proposal describe in details the proposed phased implementation approach:
+
+- [Artifact Provenance Data](#artifact-metadata)
+- [surfacing artifact metadata into the status](#from-step-into-status)
+- [surfacing the artifact metadata into the provenance payload](#from-status-into-provenance)
+
+
+### Artifact Provenance Data
+
+An artifact can be consumed or produced by a `Task`. This means that we will always have two categories at a high level: `inputs` (to indicate what was downloaded) and `outputs` (to indicate what was uploaded). The metadata content could be defined as follows:
+
+```json
+{
+    "inputs":[
+        "name": "artifact Name", # e.g. source code, dependencies
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ],
+    "outputs": [
+        "name": "artifact Name", # e.g. images, packages, coverage reports
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ]
+}
+```
+The metadata content is written by the `Step` to a file `$(step.artifacts.path)`:
+
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee $(step.artifacts.path)
+```
+The path `$(step.artifacts.path)` resolves to `/tekton/steps/<step-name>/artifacts/provenance.json` in the `Step` container at runtime.
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
+Name is a `Task`/`Step` defined name. It is useful for referencing (phase 2) and UI. The name the artifact under a particular category (`input` or `output`) must be unique.
+
+#### Values
```

Comment:
I don't think we want a list of uris here - i'd expect these to be normalized?

We do want multiple digests though so a digest can be represented by multiple hash algos

Reply:
uri + digest is a single entry. I thought we wanted a list of those for things like the `ko Task`. 

```
outputs:
  - name: images
     values:
       - uri: image1-uri
          digest: image1-digest
       - uri: image2-uri
          digest: image2-digest
       ...
```

Reply:
Do we want to support multiple digests? `e.g. sha1, sha256`
I wonder if that's a very common use case.

Reply:
Why do we need a "category" ("images", "go-packages")?

Why do we need multiple digests?

Reply:
I think I answered the concern about `category` in the above comment.  https://github.com/tektoncd/community/pull/1125/files#r1471467559

Reply:
Regarding multiple digests, I agree. Do we want to start with a single digest? If there are use cases for multiple in the future, we can add to it. I don;t know if we have a strong use case to start with multiple digests.

Reply:
Made digest into a map like we discussed at the Data Interfaces WG meeting today.